### DISCUSSION TITLE
What's the problem if I don't exclude `node_modules/` from babel-loader?

### DISCUSSION THREAD

**Discussion by nyngwang** [babel/babel#14433] [Q&A] [answered]
As title. To be specific, these are the conditions/context:

1. I don't care about speed(to a certain degree, e.g. you say I have to wait an entire day to compile then that's not acceptable to me). I want quality: smallest bundle size, and compatibility with older browsers.
2. I don't want to break something silently: it's OK that some third-party packages break because of my transpiling/polyfills, as long as the error will always get monitored.

**Comment by nicolo-ribaudo** [comment, marked as answer]
Not everything can be correctly compiled. For example, Babel rewrites async functions to promise-returning functions (via generators), and they cannot have the correct prototype because it's only accessible using async functions syntax.

If some code does this:
```js
if ((async()=>{}).constructor.name === "AsyncFunction") {
  // do something
} else {
  // do something else
}
```
your dependency will behave differently when compiled.

However, in practice Babel's transforms are "good enough" for real-world code so it's unlikely that you will find real problems.

However, make sure to:
- **not** enable `loose: true` or any [`assumptions`](babeljs.io/docs/en/assumptions) on your `node_modules` files
- set the correct `sourceType` (either by using `sourceType: "unambiguous"`, or by setting `sourceType: "script"` for all the `node_modules` files that use CommonJS), to prevent accidental introduction of strict mode by Babel.
- if you are using some ECMAScript polyfills (such as `core-js`), exclude them from your transform: they usually rely on JS quirks to detect supported features and bugs, and compiling them is likely to break something.

**Reply by nyngwang** [reply]
For the last point:

> if you are using some ECMAScript polyfills (such as `core-js`), exclude them from your transform: [...]

(**Updated**)
Does it include the case of adding `corejs` option to `@babel/preset-env`? **YES**, and don't transpile these `corejs` codes.

**Reply by nicolo-ribaudo** [reply]
You can do `exclude: /node_modules[\\/]core-js/`, to avoid compiling the polyfills you are using (i.e. `core-js`).

Starting from version ~7.2.0 (or 7.6.0, I don't remember but they are quite old anyway) it's 100% safe to compile the `@babel/*` packages.

**Reply by nyngwang** [reply]
@nicolo-ribaudo https://github.com/babel/babel/issues/9419

**Reply by nyngwang** [reply]
@nicolo-ribaudo: What's your practice? Would you mind sharing? Do you compile some of the `node_modules/` in your projects? I haven't found a reason to close this question :/ Because it seems that I haven't learned the best practice.

**Reply by nicolo-ribaudo** [reply]
I usually do not compile `node_modules`, except when I find some dependencies that need to be compiled because they are using modern syntax. The main reason is that most packages on npm are already compiled, so unnecessary re-compiling just slows down my build process.

My webpack config usually looks like this:
```js
{
  test: /\.(ts|js)x?/,
  exclude: {
    test: /node_modules/,
    not: [
      // Any dependency that needs to be compiled can be added to this array
      /module-a/,
      /module-b/
    ]
  },
  use: {
    loader: 'babel-loader',
  },
},
```

**Reply by nyngwang** [reply]
@nicolo-ribaudo: Hi, your comment is incorrect, consider an edit:

```diff
exclude: {
-    test: /node_modules/,
+    and: [/node_modules/],
     // ...
```
